File: cpp/dynamic-programming/easy/303.区域和检索-数组不可变.cpp
/*
 * @lc app=leetcode.cn id=303 lang=cpp
 *
 * [303] 区域和检索 - 数组不可变
 */

// @lc code=start
class NumArray
{
public:
    vector<int> vec_;
    vector<int> preSum_;
    NumArray(vector<int>& nums)
    {
        vec_ = nums;
        preSum_.resize(nums.size() + 1);
        for (int i = 1; i < preSum_.size(); i++) {
            preSum_[i] = preSum_[i - 1] + nums[i - 1];
        }
    }

    int sumRange(int left, int right)
    {
        return preSum_[right + 1] - preSum_[left];
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */
// @lc code=end

File: golang/prefix-sum/easy/303.区域和检索 - 数组不可变.go
/*
 * @lc app=leetcode.cn id=303 lang=golang
 * @lcpr version=30203
 *
 * [303] 区域和检索 - 数组不可变
 */

// @lc code=start
type NumArray struct {
	PreSum []int
}

func Constructor(nums []int) NumArray {
	numArray := NumArray{
		PreSum: make([]int, len(nums)+1),
	}
	for i := 1; i <= len(nums); i++ {
		numArray.PreSum[i] = numArray.PreSum[i-1] + nums[i-1]
	}
	return numArray
}

func (this *NumArray) SumRange(left int, right int) int {
	return this.PreSum[right+1] - this.PreSum[left]
}

/**
 * Your NumArray object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.SumRange(left,right);
 */
// @lc code=end

/*
// @lcpr case=start
// ["NumArray", "sumRange", "sumRange", "sumRange"]\n[[[-2, 0, 3, -5, 2, -1]], [0, 2], [2, 5], [0, 5]]\n
// @lcpr case=end

*/

